src/backend/dx12ll/src/factory.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 The Gfx-rs Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use comptr::ComPtr;
+use d3d12;
+use dxguid;
+use winapi;
+
+use std::ptr;
+use std::os::raw::c_void;
+use std::collections::BTreeMap;
+
+use core::{self, shade};
+use core::SubPass;
+use core::pso::{self, EntryPoint, GraphicsShaderSet};
+use {data, state, native};
+use {Device, Resources as R};
+
+impl core::Factory<R> for Device {
+    fn create_shader_library<'a, 'b>(&mut self, sources: shade::LibrarySource<'a, 'b>) -> Result<native::ShaderLib, shade::CreateShaderError> {
```

Comment:
Yes, although you probably see the picture better at this point. I don't think we should be baking into run-time what is known at compile time. Shader linking and loading is something very specific to the backends (at least until we get the unified language) anyway, so it wouldn't be too much harm to move `create_shader_library`  out of the trait. At the end of the day, we can consider this to be "staging" API, as in - we'll observe how it's used to make decisions on what to "stabilize" by moving it into the core trait.